### PR TITLE
Fix for count being very slow on large Parse Classes' collections (Postgres)

### PR DESCRIPTION
**Pull request by CoderickLamar**
Fetching the count for Postgres tables that are large, >5 mil in our case, takes an extremely long time that bogs down the server. This is a known issue with Postgres: https://wiki.postgresql.org/wiki/Slow_Counting

If no complex query is supplied, this will now default to estimating the count which is accurate enough for parse dashboard. 

This is an equivalent fix to commit 5264 which solved this for MongoDb.

### CODE REVIEW

**Comment by flovilmart** [comment]
That’s a great idea. Seems that a few PG tests broke along the way. Can you have a look?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by flovilmart** [comment]
>Thoughts on the broken tests? ...

@CoderickLamar we can disable this test on PG specifically if we need to (look for `it_only_db`) or we could let pg get enough time

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by flovilmart** [comment]
Usually, when a test hit the timeout, it’s because `done` is not called in time. This can be solved by handling all promises rejections with a `done.fail`. Does this make sense?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dplewis** [comment]
@CoderickLamar I looked this over real quick.

> I believe the reason this fails is because the postgres database is freshly created and hasn't had a chance to have the pg_class table built... since this is an estimate and not an exact value.

Based on the wiki you posted, the estimate value updates on certain events.

If you would have ran `analyze _Installation` or waited for the autovacuum to periodically update the estimate the test would have passed . (Note you must have autovacuum enabled on Postgres for this estimate count to work, it is defaulted to on).

I think as a fall back looking at the failing test.

```
it('query installations with count = 1', done => {
  ...
  const response = res.response;
  expect(response.results.length).toEqual(2);
  expect(response.count).toEqual(2); // returns 0 could be assign with response.results.length
  ...
});
```

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 1]
```diff
@@ -1962,15 +1962,28 @@ export class PostgresStorageAdapter implements StorageAdapter {
...
-          throw error;
-        }
-        return 0;
-      });
+    let tempQs = '';
+
+    if (where.pattern.length > 0) {
+      tempQs = `SELECT count(*) FROM $1:name ${wherePattern}`;
+    } else {
+      tempQs = `SELECT reltuples AS approximate_row_count FROM pg_class WHERE relname = '` + className + "' LIMIT 1";
```
Comment (dplewis) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Remove the templated string here and replace with 

`'SELECT reltuples AS approximate_row_count FROM pg_class WHERE relname = $1'`

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 2]
```diff
@@ -1962,15 +1962,28 @@ export class PostgresStorageAdapter implements StorageAdapter {
...
+    }
+    const qs = tempQs;
+
+    return this._client.one(qs, values, a => {
+      if(a.approximate_row_count != null){
```
Comment (dplewis) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: add spaces in if statement `if (a.approximate_row_count != null) {`

**Comment by ben-sab** [comment]
Thanks for your reply. @dplewis I have made the changes requested. As for the errors which still persist, on my local they are passing fine and they're not taking long either:

screen shot 2019-01-31 at 4 51 08 pm

Do we need to increase the timeout?

**Comment by flovilmart** [comment]
Perhaps a different Postgres version?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dplewis** [comment]
What version are running I tried it on 9.6 and 10 and it works for me locally

**Comment by flovilmart** [comment]
Travis runs 9.5

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dplewis** [comment]
@flovilmart I’ll try 9.5 but I don’t know how this would affect the schema. I’ll try tomorrow

**Comment by ben-sab** [comment]
@flovilmart @dplewis I was running PG 11 on my local.

**Comment by CoderickLamar** [comment]
@dplewis @flovilmart How can we get postgres updated on travis?

**Comment by flovilmart** [comment]
We should not, instead the fix should be compatible with Postgres 9.5.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dplewis** [comment]
I tried this on 9.5. With or without this fix I have failing test.

**Comment by flovilmart** [comment]
The tests that are failing on this branch are passing on the master branch. So the fix introduces a possible regression. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by flovilmart** [comment]
Also, excluding the tests is not something I am a big fan of, as we should maintain the same level of features on all databases 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by CoderickLamar** [comment]
@ben-sab @dplewis 
Solution here seems to be to let the estimation rebuild

Another solution would be to implement hard count specifically for deleting schema... can someone point me in the right direction for that?

**Comment by dplewis** [comment]
@behrangs Thank you for the fix. Simple and beautiful. Can you resolve the conflicts? There is a dependency bot that handles upgrading packages

**Comment by CoderickLamar** [comment]
What are the next steps here?

File: spec/InstallationsRouter.spec.js [hunk 3]
```diff
@@ -160,7 +160,7 @@ describe('InstallationsRouter', () => {
       });
   });
 
-  it('query installations with count = 1', done => {
+  it_exclude_dbs(['postgres'])('query installations with count = 1', done => {
```
Comment (dplewis) [inline spec/InstallationsRouter.spec.js]: @CoderickLamar If you remove the `it_exclude_dbs` does this still runs?

**Comment by dplewis** [comment]
@CoderickLamar @ben-sab Can you try this branch with your projects? I added the fix and a proper test for Autovalcuum.

**Comment by CoderickLamar** [comment]
Everything is looking good on my end. Speedy dashboard once again.

**Comment by CoderickLamar** [comment]
Can we get this merged into a release?

**Comment by CoderickLamar** [comment]
How often are releases usually put together on master?

**Pull request merged**